### DISCUSSION TITLE
Possibility of filters that modify a message's content

### DISCUSSION THREAD

**Discussion by equaeghe** [RealRaven2000/FiltaQuilla#310] [Q&A]
I'm seeing if I can move to Thunderbird from KMail. KMail has stability and reliability issues. [Its filtering capabilities](https://docs.kde.org/stable5/en/kmail/kmail2/filters.html) are very good, though, and I use them a lot. I'm investigating FiltaQuilla as a replacement, given that Thunderbird's built-in filtering is quite limited.

Specifically, I use KMail's filters to modify message content, such as removing headers (`Remove Header` action) and arbitrary content modification by piping the message to an external program, which then returns the modified message (`Pipe Through` action). (I use Python scripts to clean up some common issues with message content.) I've not found similar functionality in FiltaQuilla, but my understanding of the `Run Program` action is quite limited, so this is something I would want to learn more about.

So in general, I would like to learn about the possibilities for using FiltaQuilla to modify a message's content.

**Comment by RealRaven2000** [comment]
At the moment, the only thing that filters can change are _tags_ and _attachments_. Any more sophisticated reaction that generates new data would actually be relegated to my Add-on SmartTemplates, which is was augmented to collaborate with FQ to reply to or forward emails. SmartTemplates has vast potential to build new emails based on existing ones and extracting / modifying data in the "quoted" (read: original) email. It can also read all headers and incorporate them into the new mail, and [even modify headers](https://smarttemplates.quickfolders.org/variables.html#headerManipulation) (of a new email  / reply) - there is a whole language in SmartTemplates to do this. See the documentation on [premium functions](https://smarttemplates.quickfolders.org/premium.html#featureComparison).

But I haven't tried to appropriate that technology (yet) in order to try and modify _existing mail._ Not sure whether this is possible, legal or allowed. 

I haven't really looked into modifying an existing message. I think FiltaQuilla as a UI (or rather the filters dialog) is much too limited to do this is in any convenient way, and there are just no resources in an unpaid Add-on like FQ to pull this off. OTOH SmartTemplates is funded by its users and has the possibility of writing complex rules into a simple HTML (snippet) file, which probably would cover a lot of the things you had in mind.

To give you an example, I have made a macro which I use in my Support templates that clears images (of a certain minimal size) from email threads when they are quoted - a practice which is just taking up a lot of space by repeating data again and again. Since I am using inline images a ton in software support, having a template to take care of this is essential.

If we can have a discussion on modifying existing mail, I absolutely think SmartTemplates Pro would be a key component in this. But let's keep the discussion here as you also want to automate this using filters.

**Reply by equaeghe** [reply]
To be honest, I do not think Smart Templates is the solution to the problem I'm facing. _I just need a filter that hands over the message to an external program as stdin and then takes that external program's stdout and adds that to the mailstore as a message._ Whether it should replace the original message or not, is another matter; either can work. (Smart Templates's main functionality is the counterpart of what I do in my external scripts.)

**Reply by RealRaven2000** [reply]
> To be honest, I do not think Smart Templates is the solution to the problem I'm facing. _I just need a filter that hands over the message to an external program as stdin and then takes that external program's stdout and adds that to the mailstore as a message._ 

as new message? That may be possible, but you need to import the data. maybe you can generate an "eml" file and then import that somehow.

> Whether it should replace the original message or not, is another matter; either can work. (Smart Templates's main functionality is the counterpart of what I do in my external scripts.)

That;s the part where I would have looked into ST, because at least it is able to create a new message based on an existing one, and I have already done the 100s of hours of coding and testing necessary.
